contracts/integrations/test/utils/function_assertions.ts

```diff
@@ -58,11 +70,75 @@ export class FunctionAssertion {
         }
 
         // Call the after condition.
-        const afterInfo = await this.condition.after(beforeInfo, callResult, ...args);
+        const afterInfo =
+            this.condition.after !== undefined
+                ? await this.condition.after(beforeInfo, callResult, ...args)
+                : undefined;
 
         return {
             beforeInfo,
             afterInfo,
         };
     }
 }
+
+/**
+ * Note: This can be treated like a normal `FunctionAssertion`.
+ */
+export class FunctionAssertionSequence {
+    public readonly assertions: Assertion[] = [];
+    public readonly inputGenerators: Array<() => any[]> = [];
```

Comment:
let's have an `InputGenerator` type: `type InputGenerator = () => any[]`
and then have:
`public readonly assertionGenerators: Array<[Assertion, InputGenerator]>`